### DISCUSSION TITLE
curl output has single quotes when curl command is in a variable

### DISCUSSION THREAD

**Discussion by MrFreezeDZ** [curl/curl#13116] [Q&A] [answered]
We noticed that the ouptut of curl differs when the command is put into a variable.

When running ` curl -s google.de -o /dev/null -w '%{http_code}'` the output is `301`.
When running `curlcommand="curl -s google.de -o /dev/null -w '%{http_code}'"; $curlcommand` the output is `'301'`.
When running `curlcommand="curl -s google.de -o /dev/null -w '%{http_code}'"; eval $curlcommand` the ouput again is `301`.

I am just curious and want to understand why this is or if this could be a curl-bug?

Here is some output from a shell where the used bash- and curl-versions are present and the error can be seen:
```
marko_fries@cloudshell:~ $ curl -s google.de -o /dev/null -w '%{http_code}'
301marko_fries@cloudshell:~ $ curlcommand="curl -s google.de -o /dev/null -w '%{http_code}'"; $curlcommand
'301'marko_fries@cloudshell:~ $ bash --version
GNU bash, version 5.1.4(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
marko_fries@cloudshell:~ $ curl --version
curl 7.74.0 (x86_64-pc-linux-gnu) libcurl/7.74.0 OpenSSL/1.1.1w zlib/1.2.11 brotli/1.0.9 libidn2/2.3.0 libpsl/0.21.0 (+libidn2/2.3.0) libssh2/1.9.0 nghttp2/1.43.0 librtmp/2.3
Release-Date: 2020-12-09, security patched: 7.74.0-1.3+deb11u11
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
```

**Comment by icing** [comment, marked as answer]
Not sure I understand. What did you expect the `$curlcommand` to produce?

In `sh` or `bash`, using `$var` inserts the content of `var` into the command line and then chops it up to shell's quoting rules. if you want the content to be left unchanged, you'd have to use `"$curlcommand"`.

Update: ah, reading it again helps. Why are the single quotes around the second output? It is a matter how/where the shell interprets the quoting. Example:

```sh
> echo '$x'
$x
> cmd="echo '$x'"; $cmd
''
```

The single quotes *inside* the double quotes lose their special meaning. They are just a normal character there.

**Reply by MrFreezeDZ** [reply]
Hi icing,

I was expecting no difference between the curl outputs for this `curl -s google.de -o /dev/null -w '%{http_code}'` and this `curlcommand="curl -s google.de -o /dev/null -w '%{http_code}'"; $curlcommand`.

With the first command the ouptut is without quotes and with the second command the output is with single quotes.

**Reply by MrFreezeDZ** [reply]
Thank you for the update icing. Now I understand where the quotes come from.
